tools/opt/opt.cpp

```diff
@@ -112,28 +126,119 @@ NOTE: The optimizer is a work in progress.
+
+                --inline-entry-points-exhaustive
+                --eliminate-dead-code-aggressive
+
+               The following two invocations to spirv-opt are equivalent:
+
+               $ spirv-opt -Oconfig=opts.cfg program.spv
+
+               $ spirv-opt --inline-entry-points-exhaustive \
+                    --eliminate-dead-code-aggressive program.spv
+
+               Note that the flags -O, -Os and -Oconfig=<file> are also
+               affected by other flags present on the command line. For
+               example, the invocation 'spirv-opt --merge-blocks -O ...'
+               applies the transformation --merge-blocks followed by all
+               the transformations implied by -O.
+  -h, --help
                Print this help.
-  --version    
+  --version
                Display optimizer version information.
 )",
-      program, program);
+      program, program, Flags2Str(OptimizationPasses).c_str(),
+      Flags2Str(SizePasses).c_str());
 }
 
-int main(int argc, char** argv) {
-  const char* in_file = nullptr;
-  const char* out_file = nullptr;
+bool ReadFlagsFromFile(const char* oconfig_flag,
+                       std::vector<std::string>* file_flags) {
+  const char* fname = strchr(oconfig_flag, '=');
+  if (fname == nullptr || fname[0] != '=') {
+    fprintf(stderr, "error: Invalid -Oconfig flag %s\n", oconfig_flag);
+    return false;
+  }
+  fname++;
 
-  spv_target_env target_env = SPV_ENV_UNIVERSAL_1_2;
+  std::ifstream input_file;
+  input_file.open(fname);
+  if (input_file.fail()) {
+    fprintf(stderr, "error: Could not open file '%s'\n", fname);
+    return false;
+  }
 
-  spvtools::Optimizer optimizer(target_env);
-  optimizer.SetMessageConsumer([](spv_message_level_t level, const char* source,
-                                  const spv_position_t& position,
-                                  const char* message) {
-    std::cerr << StringifyMessage(level, source, position, message)
-              << std::endl;
-  });
+  while (!input_file.eof()) {
+    std::string flag;
+    input_file >> flag;
+    if (flag.length() > 0 && flag[0] != '#') {
+      file_flags->push_back(flag);
+    }
+  }
+
+  return true;
+}
+
+int ParseFlags(int argc, const char** argv, Optimizer* optimizer,
+               const char** in_file, const char** out_file);
+
+int ParseOptFlags(const char* prog_name, const char* opt_flag,
+                  Optimizer* optimizer, const char** in_file,
+                  const char** out_file) {
+  std::vector<std::string> flags;
+  flags.push_back(prog_name);
 
+  if (0 == strcmp(opt_flag, "-O")) {
+    flags.insert(flags.end(), OptimizationPasses.begin(),
+                 OptimizationPasses.end());
+  } else if (0 == strcmp(opt_flag, "-Os")) {
+    flags.insert(flags.end(), SizePasses.begin(), SizePasses.end());
+  } else if (0 == strncmp(opt_flag, "-Oconfig=", sizeof("-Oconfig=") - 1)) {
+    std::vector<std::string> file_flags;
+    if (!ReadFlagsFromFile(opt_flag, &file_flags)) {
```

Comment:
So I guess -Oconfig=foo inside a flags file does not work?   That's fine, I'm sure.  :-)

Reply:
Yeah, I was debating whether to allow it or not. I'm disallowing it for now, but we can review this later.  It's possible to have -O and -Os inside the configuration file, of course.  I figured a -Oconfig= inside a configuration file is just asking for trouble.